pkg/hubble/relay/observer/server.go

```diff
@@ -356,3 +375,44 @@ func (s *Server) ServerStatus(ctx context.Context, req *observerpb.ServerStatusR
 	}
 	return resp, g.Wait()
 }
+
+func (s *Server) heartbeat() {
+	ticker := time.NewTicker(heartbeatInterval)
+	done := make(chan bool)
+
+	s.heartbeatTicker = ticker
+	s.heartbeatDone = done
+
+	go func() {
+		for {
+			select {
+			case <-done:
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				status := s.peerConnectivityStatus()
+				s.metrics.ObserveNodesStatus(status)
+			}
+		}
+	}()
+}
+
+func (s *Server) peerConnectivityStatus() map[relaypb.NodeState]uint32 {
+	var numConnectedPeers, numUnavailablePeers uint32
+	peers := s.peers.List()
+	for _, p := range peers {
+		if !isAvailable(p.Conn) {
+			numUnavailablePeers++
+			s.opts.log.WithField("address", p.Address).Infof(
+				"No connection to peer %s, skipping", p.Name,
+			)
+			s.peers.ReportOffline(p.Name)
```

Comment:
This looks like changing some state in the peers view. As this is only used in observability/metrics code it should not affect any state and work as read-only code.